crates/sui-source-validation/src/lib.rs

```diff
@@ -41,119 +44,321 @@ pub struct BytecodeSourceVerifier<'a> {
+
+    /// If the root package needs to be verified, what address should it be fetched from?
+    fn root_address(&self, package: &CompiledPackage) -> Result<Option<AccountAddress>, Error> {
+        match self {
+            Self::Root { at: Some(addr), .. } => Ok(Some(*addr)),
+            Self::Root { at: None, .. } => Ok(Some(*package.published_at.clone()?)),
+            Self::Deps => Ok(None),
+        }
+    }
+
+    /// All the on-chain addresses that we need to fetch to build on-chain addresses.
+    fn on_chain_addresses(&self, package: &CompiledPackage) -> Result<Vec<AccountAddress>, Error> {
+        let mut addrs = vec![];
+
+        if let Some(addr) = self.root_address(package)? {
+            addrs.push(addr);
+        }
+
+        if self.verify_deps() {
+            addrs.extend(dependency_addresses(package));
+        }
+
+        Ok(addrs)
+    }
+
+    /// On-chain representation of the package and dependencies compiled to `package`, including
+    /// linkage information.
+    async fn on_chain(
         &self,
-        compiled_package: &CompiledPackage,
-    ) -> Result<(), AggregateError> {
-        self.verify_package(
-            compiled_package,
-            /* verify_deps */ true,
-            SourceMode::Skip,
+        package: &CompiledPackage,
+        verifier: &BytecodeSourceVerifier<'_>,
+    ) -> Result<OnChainRepresentation, AggregateError> {
+        let mut on_chain = OnChainRepresentation::default();
+        let mut errs: Vec<Error> = vec![];
+
+        let root = self.root_address(package)?;
+        let addrs = self.on_chain_addresses(package)?;
+
+        let resps =
+            future::join_all(addrs.iter().copied().map(|a| verifier.pkg_for_address(a))).await;
+
+        for (storage_id, pkg) in addrs.into_iter().zip(resps) {
+            let SuiRawMovePackage {
+                module_map,
+                linkage_table,
+                ..
+            } = pkg?;
+
+            let mut modules = module_map
+                .into_iter()
+                .map(|(name, bytes)| {
+                    let Ok(module) = CompiledModule::deserialize_with_defaults(&bytes) else {
+                        return Err(Error::OnChainDependencyDeserializationError {
+                            address: storage_id,
+                            module: name.into(),
+                        });
+                    };
+
+                    Ok::<_, Error>((Symbol::from(name), module))
+                })
+                .peekable();
+
+            let runtime_id = match modules.peek() {
+                Some(Ok((_, module))) => *module.self_id().address(),
+
+                Some(Err(_)) => {
+                    // SAFETY: The error type does not implement `Clone` so we need to take the
+                    // error by value. We do that by calling `next` to take the value we just
+                    // peeked, which we know is an error type.
+                    errs.push(modules.next().unwrap().unwrap_err());
```

Comment:
😓

Reply:
Is this covered by tests?

Reply:
Quite difficult to do that in this case, because it requires publishing a package that fails deserialization.